src/lib/HST/Effect/InputModule.hs

```diff
@@ -0,0 +1,143 @@
+    -- ^ A map that maps constructor names to the names of their data types.
+  }
+
+-- | An entry of the 'ModuleInterface' for a data constructor.
+data ConEntry a = ConEntry
+  { conEntryName    :: ConName a
+    -- ^ The name of the constructor.
+  , conEntryArity   :: Int
+    -- ^ The number of fields of the constructor.
+  , conEntryIsInfix :: Bool
+    -- ^ Whether the constructor should be written in infix notation or not.
+  , conEntryType    :: TypeName a
+    -- ^ The name of the data type that the constructor belongs to.
+  }
+
+-------------------------------------------------------------------------------
+-- Module Interface Utility Functions                                        --
+-------------------------------------------------------------------------------
+-- | Converts a map entry mapping a data type name to constructor entries to
+--   a list of map entries mapping the names of the given constructors to the
+--   given type name.
+revertInterfaceEntry :: (TypeName a, [ConEntry a]) -> [(ConName a, TypeName a)]
+revertInterfaceEntry (typeName, conEntry : conEntries)
+  = (conEntryName conEntry, typeName)
+  : revertInterfaceEntry (typeName, conEntries)
+revertInterfaceEntry (_, []) = []
+
+-------------------------------------------------------------------------------
+-- Effect and Actions                                                        --
+-------------------------------------------------------------------------------
+-- | An effect allowing access to the ASTs and interfaces of transformed
+--   modules by their file path. The interfaces can also be accessed by the
+--   module name.
+data InputModule a m b where
+  GetInputModule :: FilePath -> InputModule a m (S.Module a)
+  GetInputModuleInterface :: FilePath -> InputModule a m (ModuleInterface a)
+  GetInputModuleInterfaceByName :: S.ModuleName a
+    -> InputModule a m (Maybe (ModuleInterface a))
+
+makeSem ''InputModule
+
+-------------------------------------------------------------------------------
+-- Interpretations                                                           --
+-------------------------------------------------------------------------------
+-- | Handles a computation by providing the supplied module map to get modules
+--   or module interfaces from.
+--
+--   If one of the provided actions to get a module or interface by file path
+--   is used and the module is not in the map, an error is reported.
+--   If the provided action to get a module interface by module name is used
+--   and the module name is not in the map, @Nothing@ is returned.
+runInputModule :: Member Report r
+               => [(FilePath, (S.Module a, ModuleInterface a))]
+               -> Sem (InputModule a ': r) b
+               -> Sem r b
+runInputModule moduleList
+  = let moduleMap     = Map.fromList moduleList
+        moduleNameMap = Map.fromList
+          (mapMaybe createModuleNameMapEntry moduleList)
+    in interpret \case
```

Comment:
Since there is a `where`-clause, no separate `let`-expression is needed.

Reply:
I mainly use the `where` clause for local, but otherwise complete functions, whereas I use `let` expressions for calculating values that are needed more than once or for better readability.
But I can move these declarations to `where` clause if you think they fit better there. Should they get a type annotation (and maybe even a comment) like other declarations in the `where` clause?